keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -248,11 +458,833 @@ Package `kuberuntime`
+mailing list][ml-kill-pod] in the past.
+
+Another use of this functionality is that in some specific use cases, if a
+container crashes, you just want to kill the entire pod and start again. This
+was requested by some clients, at least.
+
+I'd like to know what the rest think about this.
+
+[ml-kill-pod]: https://discuss.kubernetes.io/t/how-can-i-have-a-pod-delete-itself-on-failure/8699
+
+##### Alternative 2: Do nothing
+
+Another option is to leave the pod stalled. A pod with containers that crash and
+are not restarted, will have a similar behaviour. The main and non-trivial
+difference, though, is that in this case some container will not be even
+started (non-sidecar containers won't be started if sidecars crash).
+
+
+##### Alternative 2: Always restart sidecar containers
+
+Another option is to always restart sidecar containers, under the assumptions
+that they are always needed.
+
+This seems confusing, as doing this when the pod restartPolicy is Never doesn't
+seem right. In other words, having a container not respect the pod restartPolicy
+with no explicit mentions seems asking for trouble.
+
+However, this is what the fork of at least one company is doing. Their use case
+is using Jobs to run some workloads where the main (non-sidecar) container has TB
+of memory in RAM, with no way to persist calculated data and may take weeks to run.
+In those cases losing all of that because a sidecar crashes is a too expensive
+price to pay and they resorted to always restart them.
+
+We think that in those cases using an OnFailure RestartPolicy for the job might
+seem more appropriate and, in any case, not having any kind of persistence for
+graceful shutdown doesn't seem like a good reason to have this behaviour in
+Kubernetes.
+
+In any case, it seems like a real problem and worth exploring ways to handle
+this. As always, ideas are very welcome :).
+
+##### Suggestion
+
+Go for Alternative 1. Seems simple and leaves the cluster in a clean state.
+However, I really would like to know what others think.
+
+
+#### Enforce the startup/shutdown behavior only on startup/shutdown
+
+One of the goals of this KEP is to **only modify the startup/shutdown
+sequence**. This makes the semantics clear and helps to have clean code, as only
+those places will be changed.
+
+Therefore, once non-sidecar containers have been started, kubernetes treats
+containers in the pod indistinguishable from containers in a pod without
+sidecar containers. That is, once sidecars and non-sidecars containers have
+been started once, the regular kubernetes reconciliation loop is used for all
+the containers in the pod during the pod lifecycle until the shutdown sequence
+is fired. This guarantees that only the startup and shutdown behaviour of
+sidecar containers is affected and not the rest of the pod lifecycle.
+
+However, one side effect is that if, for example, all containers happen to
+crash at the same time, all _can_ be restarted at the same time (if the restart
```

Comment:
Having them restart using the same startup/shutdown behavior would allow providing the guarantee to the main container that the sidecar will always be there regardless of the circumstances, which would be very valuable. I understand however if this can complicate the implementation and thus be regarded as out of scope :wink:

Reply:
I think that is not really possible, there always be races:
 * sidecar A is running and container B is running
 * container B crashes, is being restarted normally (as A is still running)
 * Before B is ready or fully started, A crashes
 * B starts without A running

I think there is no way to prevent those races, so it seems just simpler to not try to hide something from the user, IMHO, and make it clear that those things can happen.

What do you think?

Reply:
Agreed, it can get very complex pretty quickly!

Reply:
Coo, will mark it as solved, then! Thanks for the feedback :)

Reply:
btw, I don't think it is _possible_ not just _complicated_. If you think it will be useful, I can add a note on that